### PR TITLE
Initial FF1R implementation

### PR DESCRIPTION
**Pull request by jtoyoda**
Potential future improvements:
- Add in obfuscation for the yaml
- Add in a weapon/armor queue in the client
- Add in death link support
- Add in web tracker

### CODE REVIEW

File: data/lua/FF1/socket.lua [hunk 1]
```diff
@@ -0,0 +1,132 @@
+-----------------------------------------------------------------------------
```
Comment (jtoyoda) [inline data/lua/FF1/socket.lua:1]: https://github.com/diegonehab/luasocket/blob/master/LICENSE

File: data/lua/FF1/json.lua [hunk 2]
```diff
@@ -0,0 +1,380 @@
+--
+-- json.lua
+--
+-- Copyright (c) 2015 rxi
```
Comment (jtoyoda) [inline data/lua/FF1/json.lua:4]: https://github.com/rxi/json.lua/blob/master/LICENSE

File: FF1Client.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+                continue
+            except ConnectionRefusedError:
+                logger.debug("Connection Refused, Trying Again")
+                ctx.nes_status = CONNECTION_REFUSED_STATUS
+                continue
+
+
+if __name__ == '__main__':
+    # Text Mode to use !hint and such with games that have no text entry
+    Utils.init_logging("TextClient")
```
Comment (Berserker66) [inline FF1Client.py:214]: You should do FF1Client instead of TextClient here.

File: FF1Client.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        if ctx.server_task:
+            await ctx.server_task
+        if ctx.nes_sync_task:
+            await ctx.nes_sync_task
+
+        while ctx.input_requests > 0:
+            ctx.input_queue.put_nowait(None)
+            ctx.input_requests -= 1
+
+        if ui_task:
```
Comment (Berserker66) [inline FF1Client.py:243]: there is now ctx.shutdown(), so you can remove a few lines here if you want.

File: FF1Client.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            await ui_task
+
+        if input_task:
+            input_task.cancel()
+
+
+    import argparse
+    import colorama
+
+    parser = argparse.ArgumentParser(description="FF1 Archipelago Client")
```
Comment (Berserker66) [inline FF1Client.py:253]: CommonClient.get_base_parser should be preferred here.

File: WebHostLib/static/assets/tutorial/ff1/multiworld_en.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
+# Final Fantasy 1 (NES) Multiworld Setup Guide
+
+## Required Software
+- The FF1Client which is bundled with [Archipelago](https://github.com/ArchipelagoMW/Archipelago/releases)
+- The [BizHawk](http://tasvideos.org/BizHawk.html) emulator. Versions 2.3.1 and higher are supported. 
+  Version 2.7 is recommended
+- Your Final Fantasy (USA Edition) ROM file, probably named `Final Fantasy (USA).nes`. Archipelago.gg nor the Final
```
Comment (Berserker66) [inline WebHostLib/static/assets/tutorial/ff1/multiworld_en.md:7]: I think "Neither Archipelago nor the [...]" is better englissh, right?
(Not actually a native English speaker, so I definitely can be wrong.)

File: FF1Client.py (hunk 4 above)
Comment (jtoyoda) [reply FF1Client.py:243]: The issue is that I am awaiting nes_sync_task as well. I am not sure how that should be handled with ctx.shutdown. I could just call it after but that felt weird as well

File: WebHostLib/static/assets/tutorial/ff1/multiworld_en.md (hunk 6 above)
Comment (jtoyoda) [reply WebHostLib/static/assets/tutorial/ff1/multiworld_en.md:7]: I mean I am but I don't know either. "Neither" feels correct here

File: FF1Client.py (hunk 4 above)
Comment (Berserker66) [reply FF1Client.py:244]: look at FactorioClient.main, it's just to consolidate some of the ctx shutdowns. Of course you could extend it and then call super() in your method, whichever way you prefer.

**Comment by Berserker66** [comment]
Alright, I think that looks good, waiting for that hopefully final sunday test game.

File: FF1Client.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    def on_package(self, cmd: str, args: dict):
+        if cmd == 'Connected':
+            self.game = self.games.get(self.slot, None)
+            asyncio.create_task(parse_locations(self.locations_array, self, True))
+        elif cmd == 'Print':
+            msg = args['text']
+            if ': !' not in msg:
+                self._set_message(msg, SYSTEM_MESSAGE_ID)
+        elif cmd == "ReceivedItems":
+            msg = f"Recieved {', '.join([self.item_name_getter(item.item) for item in args['items']])}"
```
Comment (Berserker66) [inline FF1Client.py:67]: Typo, i and e reversed

**Pull request merged**